### PR TITLE
gh-99110: Initialize frame->previous in init_frame to fix segmentation fault

### LINKED ISSUE #99110: Segfault on frame.f_back when frame is created with PyFrame_New()
Python segfaults when frame.f_back is accessed on a frame created with PyFrame_New() c api. Calling the PyFrame_GetBack() c api also segfaults, at least in debug builds and on win32 (it depends on the contents of uninitialized memory). Tested with 3.11.0 and git 3.11 branch as of Nov 4, 2022

Cause is that the ->previous field of the _PyInterpreterFrame is never set to NULL and when PyFrame_GetBack() runs, it tries to dereference the pointer value of ->previous and segfaults. A test case using ctypes is attached.

Adding a frame->previous = NULL; line to init_frame() in frameobject.c fixes this, though I don't know if it's the best place for it.

[f_back_segfault.py.txt](https://github.com/python/cpython/files/9942212/f_back_segfault.py.txt)

### Linked PRs
* gh-100182
* gh-100478

### PR DESCRIPTION
**Pull request by byllyfish**
This PR fixes a segmentation fault caused by an uninitialized variable. One way to trigger the seg. fault is to use the PyFrame_New() C API. I've also run across this segmentation fault by calling `asyncio.Task.get_stack()` on cython based asyncio Task (in the grpcio module).

The fix for #99110 needs to be back-ported to 3.11.

* Issue: gh-99110

### CODE REVIEW

File: Lib/test/test_ctypes/test_python_api.py [hunk 1]
```diff
@@ -81,5 +81,36 @@ def test_pyobject_repr(self):
...
+
+        frame = pythonapi.PyFrame_New(
+            pythonapi.PyThreadState_Get(),          # thread state
+            cast(id(dummy.__code__), p_memtype),    # a code object
+            globals(),
+            locals(),
+        )
+
+        # The following line should not cause a segmentation fault.
+        assert frame.f_back is None
```
Comment (kumaraditya303) [inline Lib/test/test_ctypes/test_python_api.py]: Use self.assertEqual

**Comment by markshannon** [comment]
`_PyInterpreterFrame.previous` gets initialized when the frame is linked into the frame stack.
I'd rather not add extra code to `_PyFrame_InitializeSpecials`, as it is performance critical.

`_PyInterpreterFrame.previous` could be set to `NULL` [when initializing the frame object](https://github.com/python/cpython/blob/main/Objects/frameobject.c#L1016)

File: Include/internal/pycore_frame.h [hunk 2]
```diff
@@ -112,6 +112,7 @@ _PyFrame_InitializeSpecials(
     frame->prev_instr = _PyCode_CODE(code) - 1;
     frame->yield_offset = 0;
     frame->owner = FRAME_OWNED_BY_THREAD;
+    // frame->previous: initialized when frame is linked into the frame stack
```
Comment (markshannon) [inline Include/internal/pycore_frame.h]: This looks a bit like commented out code.
Could you move the comment to above the function.

Comment (markshannon) [reply Include/internal/pycore_frame.h]: Maybe change `/* Consumes reference to func and locals */` to 
```C
/* Consumes reference to func and locals.
   Does not initialize frame->previous, which happens
   when frame is linked into the frame stack.
*/
```

File: Lib/test/test_ctypes/test_python_api.py [hunk 3]
```diff
@@ -81,5 +81,36 @@ def test_pyobject_repr(self):
         self.assertEqual(repr(py_object(42)), "py_object(42)")
         self.assertEqual(repr(py_object(object)), "py_object(%r)" % object)
 
+    def test_PyFrame_New_f_back(self):
```
Comment (markshannon) [inline Lib/test/test_ctypes/test_python_api.py]: This sort of test usually goes in `test_capi`.

First add a function to call `PyFrame_New` in `_testcapimodule.c`, then call that function in the test.

So, `pythonapi.PyFrame_New(...)` below would be replaced with `_testcapimodule.new_frame(dummy.__code__, globals(), locals())`.

**Review by markshannon — CHANGES_REQUESTED** [review]

**Comment by byllyfish** [comment]
I moved the test code into the `TestCAPI` unit test class in test_frame.py.

I have made the requested changes; please review again.

**Review by kumaraditya303 — APPROVED** [review]

File: Lib/test/test_frame.py [hunk 4]
```diff
@@ -408,6 +408,15 @@ def test_frame_get_generator(self):
...
 
+    def test_frame_fback_api(self):
+        """Test that accessing `f_back` does not cause a segmentation fault on
+        a frame created with `PyFrame_New` (GH-99110)."""
+        def dummy():
+            pass
+
+        frame = _testcapi.frame_new(dummy.__code__, globals(), locals())
+        # The following line should not cause a segmentation fault.
+        self.assertEqual(frame.f_back, None)
```
Comment (kumaraditya303) [inline Lib/test/test_frame.py]:
```suggestion
        self.assertIsNone(frame.f_back)
```

**Review by markshannon — APPROVED** [review]
Looks good now. Thanks.

File: Misc/NEWS.d/next/Core and Builtins/2022-12-12-01-05-16.gh-issue-99110.1JqtIg.rst [hunk 5]
```diff
@@ -0,0 +1,2 @@
+Initialize frame->previous in frameobject.c to fix a segmentation fault when
+accessing frames created by ctypes.
```
Comment (kumaraditya303) [inline Misc/NEWS.d/next/Core and Builtins/2022-12-12-01-05-16.gh-issue-99110.1JqtIg.rst]:
```suggestion
accessing frames created by :c:func:`PyFrame_New`.
```

**Pull request merged**